src/toolbox/xor.rs

```diff
@@ -0,0 +1,270 @@
+
+        return Ok(shares);
+    }
+
+    /// For Xor, sparse_shares should include a None value for each share you want us to generate
+    fn complete(&mut self, secret: &Scalar, sparse_shares: &Vec<Option<Scalar>>) -> Result<Vec<Scalar>, String> {
+        let mut empties = 0;
+        let mut shares = Vec::new();
+        let mut new_shares = Vec::new();
+        for share in sparse_shares.iter() {
+            if share.is_some() {
+                shares.push(share.unwrap());
+            } else {
+                if empties != 0 {     // the "zero-th" empty will be the one we have to XOR-calculate
+                    trace!("Pushing a random Scalar to extra empty");
+                    new_shares.push(Scalar::random(&mut self.rng));
+                }
+                empties += 1;
+            }
+        }
+
+        if empties < 1 {
+            return Err(String::from("No empty shares in which to complete!"));
+        }
+
+        info!("Completing {} empties with {} known and secret = {:?}", empties, shares.len(), secret);
+
+        match xor_many_more_scalars(secret, &shares, &new_shares) {
+            Ok(val) => {
+                debug!("The XOR'd share is {:?}", val);
+                new_shares.push(val);
+            },
+            Err(e) => return Err(e),
+        }
+
+        let ret_shares: Vec<Scalar> = sparse_shares.iter().map(|x| {
+            if x.is_some() {
+                x.unwrap()
+            } else {
+                // it's OK to panic if we get a None here, since that should *never* happen
+                new_shares.pop().unwrap()
+            }
+        }).collect();
+
+        return Ok(ret_shares);
+    }
+
+    fn reconstruct(&mut self, shares: &Vec<Scalar>) -> Result<Scalar, String> {
+        if shares.len() < 1 {
+            return Err(String::from("No shares provided, impossible to reconstruct!"));
+        }
+
+        let secret = xor_many_scalars(&shares[0], &shares[1..]);
+        
+        debug!("Used {} shares to reconstruct {:?}", shares.len(), secret);
+        return secret;
+    }
+}
+
+/// Given two Scalar values, return the XOR of their constituent byte arrays
+pub fn xor_scalars(a: &Scalar, b: &Scalar) -> Scalar {
+    let mut the_bytes = [0 as u8; 32];
+    let abytes = a.to_bytes();
+    let bbytes = b.to_bytes();
+    for i in 0..the_bytes.len() {
+        the_bytes[i] = abytes[i] ^ bbytes[i];
+    }
+    Scalar::from_bits(the_bytes)
+}
+
+/// Given a list of scalars, XOR them all together.  We split out the `first` separately for efficiency's sake;
+/// there are use cases where the initial value you want to XOR with isn't in your Vector, so you don't need to
+/// push it.
+/// 
+/// TODO WARNING: the result isn't guaranteed to be modulo group order.  Is that going to be a problem?
+/// TODO is there a special way we need to write this function, to achieve some sort of SIMD benefit?
+pub fn xor_many_scalars(first: &Scalar, others: &[Scalar]) -> Result<Scalar, String> {
```

Comment:
Why not just 1 parameter that is an array of `Scalar`?

Reply:
The different function signatures make calling this code from `impl SecretSharing` easier, because it fits with the variables we have at the time.  It lets us avoid creating new Vectors just so we can hand them over to be XOR'd.  I don't have the functions call each other because that uses memory we don't need.  In each version, there's just a single `u8` array which takes all the changes, which should be the fastest possible way to run this code.

I'm still getting a feel for this group's specific code dupe vs. performance vs. readability lines, so I'm happy to revisit if this doesn't fit where you want it to.